blob/drivertest/drivertest.go

```diff
@@ -134,6 +140,169 @@ func RunConformanceTests(t *testing.T, newHarness HarnessMaker, asTests []AsTest
+	// TODO(Issue #541): Add tests for slash-separated paths.
+	const keyPrefix = "blob-for-list"
+	content := []byte("hello")
+
+	keyForIndex := func(i int) string { return fmt.Sprintf("%s-%d", keyPrefix, i) }
+	indexFromKey := func(key string) (int, error) {
+		if !strings.HasPrefix(key, keyPrefix) {
+			return 0, fmt.Errorf("got name %q, expected it to have prefix %q", key, keyPrefix)
+		}
+		return strconv.Atoi(key[len(keyPrefix)+1:])
+	}
+
+	tests := []struct {
+		name       string
+		skipCreate bool
+		pageSize   int
+		prefix     string
+		want       [][]int
+		wantIter   []int
+		wantErr    bool
+	}{
+		{
+			name:       "negative page size returns an error",
+			skipCreate: true,
+			pageSize:   -1,
+			wantErr:    true,
+		},
+		{
+			name:       "page size out of range returns an error",
+			skipCreate: true,
+			pageSize:   blob.MaxPageSize + 1,
+			wantErr:    true,
+		},
+		{
+			name:   "no objects",
+			prefix: "no-objects-with-this-prefix",
+			want:   [][]int{nil},
+		},
+		{
+			name:     "exactly 1 object due to prefix",
+			prefix:   keyForIndex(1),
+			want:     [][]int{[]int{1}},
+			wantIter: []int{1},
+		},
+		{
+			name:     "no pagination",
+			prefix:   keyPrefix,
+			want:     [][]int{[]int{0, 1, 2}},
+			wantIter: []int{0, 1, 2},
+		},
+		{
+			name:     "by 1",
+			prefix:   keyPrefix,
+			pageSize: 1,
+			want:     [][]int{[]int{0}, []int{1}, []int{2}},
+			wantIter: []int{0, 1, 2},
+		},
+		{
+			name:     "by 2",
+			prefix:   keyPrefix,
+			pageSize: 2,
+			want:     [][]int{[]int{0, 1}, []int{2}},
+			wantIter: []int{0, 1, 2},
+		},
+		{
+			name:     "by 3",
+			prefix:   keyPrefix,
+			pageSize: 3,
+			want:     [][]int{[]int{0, 1, 2}},
+			wantIter: []int{0, 1, 2},
+		},
+	}
+
+	ctx := context.Background()
+
+	// Creates blobs for sub-tests below.
+	// We only create the blobs once, for efficiency and because there's
+	// no guarantee that after we create them they will be immediately returned
+	// from List/ListPaged. The very first time the test is run against a
+	// Bucket, it may be flaky due to this race.
+	init := func(t *testing.T, skipCreate bool) (*blob.Bucket, func()) {
+		h, err := newHarness(ctx, t)
+		if err != nil {
+			t.Fatal(err)
+		}
+		b, err := h.MakeBucket(ctx)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !skipCreate {
+			// See if the blobs are already there.
+			iter, err := b.List(ctx, &blob.ListOptions{Prefix: keyPrefix})
+			if err != nil {
+				t.Fatal(err)
+			}
+			count := 0
+			for {
+				obj, err := iter.Next(ctx)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if obj == nil {
+					break
+				}
+				count++
+			}
+			if count != 3 {
+				for i := 0; i < 3; i++ {
+					if err := b.WriteAll(ctx, keyForIndex(i), content, nil); err != nil {
+						t.Fatal(err)
+					}
+				}
+			}
+		}
+		return b, func() { h.Close() }
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			b, done := init(t, tc.skipCreate)
```

Comment:
No, we can't call `init` only once, because we need a new harness and Bucket for each test (so that they end up with separate `.yaml` files). We could theoretically only do the "do the blobs exist and if not create them" only for the first test, but that would break if you used `--run` to only run a single test (because only the first test has the `List` recorded in it, and if you skip some tests then the first one changes).

We could potentially run more tests in parellel; I think that's true across this file. The recorded tests are quite fast (< 0.5s) so I'm not sure it's worth it, but if you think it's a good idea you could file a separate issue.